### PR TITLE
feat: Caption services (608/708) metadata

### PR DESCRIPTION
**Pull request by gkatsev**
Add an option for caption services metadata in case the user wants to specify labels for 608/708 captions, override the ones provided in the manifest, or needs to add more information like character encoding (this isn't currently available but will be added some time in the future).
For HLS, an `EXT-X-MEDIA` tag can be specified with an `INSTREAM-ID` attribute. We already support this. https://datatracker.ietf.org/doc/html/draft-pantos-hls-rfc8216bis-08#section-4.4.6.1
This PR updated mpd-parser which uses the ANSI 214 supplamental spec section 7.2 to parse out the same information from MPD files. https://github.com/videojs/mpd-parser/pull/131.

Adds a property called `captionServices` which has properties of the caption service IDs like `CC1` or `SERVICE1` and allows a user to specify a `language` and `label`.

### CODE REVIEW

File: src/util/text-tracks.js [hunk 1]
```diff
@@ -24,12 +24,28 @@ export const createCaptionsTrackIfNotExists = function(inbandTextTracks, tech, c
       // in the m3u8 for us to use
       inbandTextTracks[captionStream] = track;
     } else {
+      const captionServices = tech.options_.vhs && tech.options_.vhs.captionServices || {};
```
Comment (gkatsev) [inline src/util/text-tracks.js:27]: This section gets used when we have caption services, but they aren't being specified in the manifests. When specified in the manifest, we will apply the properties in media groups https://github.com/videojs/http-streaming/blob/ad9aa2ef67fa90557af6afa7878d911b5f1093ac/src/media-groups.js#L641-L662
which will need an update to use this captionServices data

File: src/util/text-tracks.js [hunk 2]
```diff
@@ -16,20 +16,41 @@ export const createCaptionsTrackIfNotExists = function(inbandTextTracks, tech, c
...
+      }
+
       // Otherwise, create a track with the default `CC#` label and
       // without a language
       inbandTextTracks[captionStream] = tech.addRemoteTextTrack({
         kind: 'captions',
-        id: captionStream,
-        label: captionStream
+        id: instreamId,
+        default: def,
```
Comment (gkatsev) [inline src/util/text-tracks.js:51]: For whatever reason this doesn't seem to do anything, though, I'd expect it to. It does let us override it above in the media groups sections.

File: README.md [hunk 3]
```diff
@@ -466,6 +469,47 @@ This option defaults to `false`.
 * Default: [`SAFE_TIME_DELTA`](https://github.com/videojs/http-streaming/blob/e7cb63af010779108336eddb5c8fd138d6390e95/src/ranges.js#L17)
 * Allow to  re-define length (in seconds) of time delta when you compare current time and the end of the buffered range.
 
+##### captionServices
+* Type: `object`
+* Default: undefined
+* Allow to override instream captions like 608 and 708 to provide labels or languaages.
```
Comment (gesinger) [inline README.md:475]:
```suggestion
* Allow overriding of instream captions like 608 and 708 to provide labels and/or languages.
```

File: src/util/text-tracks.js [hunk 4]
```diff
@@ -16,20 +16,41 @@ export const createCaptionsTrackIfNotExists = function(inbandTextTracks, tech, c
...
+      const captionServices = tech.options_.vhs && tech.options_.vhs.captionServices || {};
+      let label = captionStream;
+      let language = captionStream;
+      let def = false;
+
+      if (captionServices[instreamId]) {
+        label = captionServices[instreamId].label;
+        language = captionServices[instreamId].language;
+        def = captionServices[instreamId].default;
+      }
```
Comment (gesinger) [inline src/util/text-tracks.js:44]:
```suggestion
      const captionService = captionServices[instreamId];

      if (captionService) {
        label = captionService.label;
        language = captionService.language;
        def = captionService.default;
      }
```

File: src/util/text-tracks.js (hunk 1 above)
Comment (gesinger) [reply src/util/text-tracks.js:35]: Might be worth adding a comment here mentioning that.

File: src/util/text-tracks.js (hunk 2 above)
Comment (gesinger) [reply src/util/text-tracks.js:51]: Worth a comment to look into it?

File: test/master-playlist-controller.test.js [hunk 5]
```diff
@@ -3260,15 +3260,77 @@ QUnit.test(
       .map(cap => Object.assign({name: cap.id}, cap));
 
     assert.equal(capsArr.length, 4, '4 closed-caption tracks defined in playlist');
-    assert.equal(addedCaps.length, 2, '2 CEA608 tracks added internally');
+    assert.equal(addedCaps.length, 4, '4 tracks added internally');
```
Comment (gesinger) [inline test/master-playlist-controller.test.js:3263]:
```suggestion
    // 2 608 tracks, 2 708 tracks
    assert.equal(addedCaps.length, 4, '4 tracks added internally');
```

File: test/master-playlist-controller.test.js [hunk 6]
```diff
@@ -3260,15 +3260,77 @@ QUnit.test(
...
+    // media
+    this.standardXHRResponse(this.requests.shift());
+
+    const master = masterPlaylistController.masterPlaylistLoader_.master;
+    const caps = master.mediaGroups['CLOSED-CAPTIONS'].CCs;
+    const capsArr = Object.keys(caps).map(key => Object.assign({name: key}, caps[key]));
+    const addedCaps = masterPlaylistController.mediaTypes_['CLOSED-CAPTIONS'].groups.CCs
+      .map(cap => Object.assign({name: cap.id}, cap));
+
+    assert.equal(capsArr.length, 4, '4 closed-caption tracks defined in playlist');
```
Comment (gesinger) [inline test/master-playlist-controller.test.js:3327]:
```suggestion
    // 2 608 tracks, 2 708 tracks
    assert.equal(capsArr.length, 4, '4 closed-caption tracks defined in playlist');
```

Comment (gkatsev) [reply test/master-playlist-controller.test.js:3327]: Added it as part of the test message

File: test/master-playlist-controller.test.js (hunk 5 above)
Comment (gkatsev) [reply test/master-playlist-controller.test.js]: Added it as part of the test message

File: README.md (hunk 3 above)
Comment (gkatsev) [reply README.md]: Reworded entirely.

File: src/util/text-tracks.js (hunk 2 above)
Comment (gkatsev) [reply src/util/text-tracks.js:55]: added

File: src/util/text-tracks.js (hunk 1 above)
Comment (gkatsev) [reply src/util/text-tracks.js:37]: Yeah, makes sense. Added

**Review by gesinger — APPROVED** [review]

**Pull request merged**